src/main/java/net/neoforged/neoforge/common/util/INBTSerializable.java

```diff
@@ -5,15 +5,51 @@
 
 package net.neoforged.neoforge.common.util;
 
+import com.mojang.serialization.Codec;
+import java.util.function.Supplier;
 import net.minecraft.core.HolderLookup;
+import net.minecraft.nbt.NbtAccounter;
 import net.minecraft.nbt.Tag;
+import net.minecraft.network.RegistryFriendlyByteBuf;
+import net.minecraft.network.codec.StreamCodec;
+import net.neoforged.neoforge.codec.NBTSerializableCodec;
 import org.jetbrains.annotations.UnknownNullability;
 
 /**
  * An interface designed to unify various things in the Minecraft
  * code base that can be serialized to and from a NBT tag.
  */
 public interface INBTSerializable<T extends Tag> {
+    /**
+     * Creates a codec for an implementation of {@link INBTSerializable}.
+     *
+     * @param serializableClass The class implementing INBTSerializable.
+     * @param defaultSupplier   A default supplier for the class; typically a method reference.
+     * @param <S>               Type of class implementing INBTSerializable.
+     * @param <NBT>             Type of NBT tag the class serializes to.
+     * @return A codec for the serializable class.
+     */
+    static <S extends INBTSerializable<NBT>, NBT extends Tag> Codec<S> codec(Class<S> serializableClass, Supplier<S> defaultSupplier) {
```

Comment:
As noted above, delegating everything off the class this way just seems like the wrong approach at a fundamental level.